### PR TITLE
Converting custom ITelemetry implementations into Event Telemetry with properties

### PR DESCRIPTION
**Pull request by Dmitry-Matveev**
Addressing enhancement #988.
Initial implementation of `DictionarySerializationWriter` to write a `Dictionary<string, string>` for all serialized fields in the given `ITelemetry`:
- Pending some perf optimizations;
- Note: most places with couple string concatenations are "+", more than couple is `Concat()`, no serial string concatenations to benefit from `StringBuilder`.
- Not sure on benefits of initializing `Stack` for each serialized item - something to think about...

- [x] I ran [Unit Tests](https://github.com/Microsoft/ApplicationInsights-dotnet/blob/develop/.github/CONTRIBUTING.md) locally.

For significant contributions please make sure you have completed the following items:

- [ ] Design discussion issue #988
- [x] Changes in public surface reviewed
- [x] CHANGELOG.md updated with one line description of the fix, and a link to the original issue.
- [x] The PR will trigger build, unit test, and functional tests automatically. If your PR was submitted from fork - mention one of committers to initiate the build for you.

### CODE REVIEW

File: Test/Microsoft.ApplicationInsights.Test/Shared/Extensibility/Implementation/IExtensionImplementations.cs [hunk 1]
```diff
@@ -0,0 +1,83 @@
+﻿namespace Microsoft.ApplicationInsights.Extensibility.Implementation
+{
+    using System;
+    using System.Linq;
+    using System.Collections.Generic;
+
+    public class MyTestExtension : IExtension
```
Comment (Dmitry-Matveev) [inline Test/Microsoft.ApplicationInsights.Test/Shared/Extensibility/Implementation/IExtensionImplementations.cs]: Moved several existing similar test classes into one place..

File: src/Microsoft.ApplicationInsights/Extensibility/Implementation/DictionarySerializationWriter.cs [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+
+            string key = this.GetKey(name);
+
+            if (!this.AccumulatedDictionary.ContainsKey(key))
+            {
+                this.AccumulatedDictionary[key] = value;
+            }
+        }
+
+        public void WriteProperty(string name, double? value)
```
Comment (Dmitry-Matveev) [inline src/Microsoft.ApplicationInsights/Extensibility/Implementation/DictionarySerializationWriter.cs]: We may want to consider this as a metric, but there are several places and use-cases where it should be property but not metric... Storing it as Property for the time being.

**Comment by MS-TimothyMothra** [comment]
Can we discuss this?
This might not be necessary with a simple modification to `ITelemetry`

**Comment by Dmitry-Matveev** [comment]
Discussed offline with @MS-TimothyMothra, PR is still necessary :) :) :)

File: src/Microsoft.ApplicationInsights/Extensibility/Implementation/DictionarySerializationWriter.cs [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+                    this.AccumulatedDictionary[key] = value.Value.ToString(CultureInfo.InvariantCulture);
+                }
+                else
+                {
+                    this.AccumulatedDictionary[key] = null;
+                }
+            }
+        }
+
+        public void WriteProperty(string name, int? value)
```
Comment (MS-TimothyMothra) [inline src/Microsoft.ApplicationInsights/Extensibility/Implementation/DictionarySerializationWriter.cs]: I see that this method is copied several times based on the input parameter. could this accept a generic Nullable parameter?

Comment (Dmitry-Matveev) [reply src/Microsoft.ApplicationInsights/Extensibility/Implementation/DictionarySerializationWriter.cs]: This is an implementation of `ISerializationWriter`, I would not be able to change this on implementing side. @cijothomas would know if we can change it on the Interface side. My expectation that we would not, it already requires a conditional compilation on `bool?` between .NET and .NET Core, so not all Nullables are treated the same.

File: src/Microsoft.ApplicationInsights/Extensibility/Implementation/DictionarySerializationWriter.cs [hunk 4]
```diff
@@ -0,0 +1,327 @@
+﻿namespace Microsoft.ApplicationInsights.Extensibility.Implementation
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Globalization;
+
+    internal class DictionarySerializationWriter : ISerializationWriter
```
Comment (cijothomas) [inline src/Microsoft.ApplicationInsights/Extensibility/Implementation/DictionarySerializationWriter.cs]: comment the usage of this class for future devs :)

**Pull request merged**